ServerCore/Pages/Teams/CreateTeam.razor.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Forms;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using ServerCore.DataModel;
+using ServerCore.Helpers;
+using ServerCore.ModelBases;
+using System.Linq;
+
+namespace ServerCore.Pages.Teams
+{
+    class EmailListValidationAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            string emailList = value as string;
+            if (string.IsNullOrWhiteSpace(emailList))
+            {
+                return new ValidationResult("An email is required.");
+            }
+
+            if (!MailHelper.IsValidEmail(emailList))
+            {
+                return new ValidationResult("Invalid primary contact email.");
+            }
+
+            return ValidationResult.Success;
+        }
+    }
+
+    class TeamUniquenessValidationAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            PuzzleServerContext puzzleServerContext = validationContext.GetRequiredService<PuzzleServerContext>();
+            TeamModel teamModel = (TeamModel)validationContext.ObjectInstance;
+            string name = value as string;
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                return new ValidationResult("Team names cannot be left blank.");
+            }
+
+            if (TeamHelper.IsTeamNameTaken(puzzleServerContext, teamModel.EventId, name))
```

Comment:
```Wait()``` deadlocks here and I couldn't find a way to prevent it. Normally ```.ConfigureAwait(false)``` would work around that kind of deadlock, but somehow Blazor validation gets stuck anyway. We could preload team names, but that risks a team being created with that name sometime after the page loads. I didn't think synchronous call would be too bad here since creating teams is a rare operation, but what do you think?

Reply:
Understood - can we have two versions then, one of which is synchronous and is used only in this attribute, and one which is async and used everywhere else? It should be straightforward to put the truncation-check logic into a private helper function used by both.